docs/design/resource-qos.md

```diff
@@ -0,0 +1,220 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/design/resource-qos.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Resource Quality of Service in Kubernetes
+
+**Author**: Ananya Kumar (@AnanyaKumar)
+
+**Status**: Implemented
+
+*This document presents the design of resource quality of service for containers in Kubernetes, and describes use cases and implementation details.*
+
+## Motivation
```

Comment:
This section made sense when it was written but is no longer accurate. (It's basically saying "today Kubernetes just uses limit, here's why that's bad, and here's how we're going to make it better") 

I would rewrite as follows:

In the first version of Kubernetes, users specified one value, a resource limit, for each resource of a container. For example, a user might specify 1GB memory limit. The scheduler used resource limits to schedule containers (technically, the scheduler schedules pods comprised of containers). For example, the scheduler would not place 5 containers each with a 1GB memory limit onto a machine with 4GB of memory.

This had the drawback that if users over-estimate the resource needs of their containers, then cluster utilization will be poor, since the unused resources are stranded and can't be used by other containers. On the other hand, if they under-estimate the resource needs (specifying no limits is a special case of under-estimating), then the machine could become overcommitted, leading to out-of-memory or CPU starvation of **any** container on the node, even the ones that specified accurate resource limits.

This document describes the way Kubernetes provides different levels of Quality of Service to containers depending on what they request. Containers that need to stay up reliably can request guaranteed resources, while containers with less stringent requirements can use resources with weaker or no guarantee. Specifically, for each resource, containers specify a *request*, which is the amount of that resource that the system will guarantee to the container, and a *limit* which is the maximum amount that the system will allow the container to use. When request == limit, the resources are guranteed, and when request \< limit, the container is guaranteed the request but can opportunistically scavenge the difference between request and limit if they are not being used by other containers. This allows Kubernetes to oversubscribe nodes, which increases utilization, while at the same time maintaining resource guarantees for the containers that need guarantees. Borg increased utilization by about 20% when it started allowing use of such non-guaranteed resources, and we hope to see similar improvements in Kubernetes.